models/intermediate/int_zendesk__schedule_spine.sql

```diff
@@ -90,21 +95,270 @@ with timezone as (
+    from holiday_check
+    {{ dbt_utils.group_by(n=9) }}
+
+-- Since we have holiday schedules and normal schedules, we need to union them into a holistic schedule spine
+), spine_union as (
+
+    select
+        schedule_id, 
+        time_zone, 
+        schedule_name, 
+        valid_from, 
+        valid_until, 
+        start_time_utc, 
+        end_time_utc, 
+        holiday_week_start,
+        holiday_week_end,
+        holiday_name_check
+    from holiday_consolidated
+
+    union all
+
+    select
+        schedule_id, 
+        time_zone, 
+        schedule_name, 
+        valid_from, 
+        valid_until, 
+        start_time_utc, 
+        end_time_utc, 
+        null as holiday_week_start,
+        null as holiday_week_end,
+        null as holiday_name_check
     from calculate_schedules
+
+-- Within the spine, determine which schedules contain a holiday
+), valid_adjustment as (
+
+    select 
+        *,
+        case when max(holiday_name_check) over (partition by schedule_id, holiday_week_start order by holiday_week_start rows unbounded preceding) is not null then 1 else 0 end as is_holiday_week
+    from spine_union
+
+-- Now that we have an understanding of which weeks are holiday's let's consolidate them with non holiday weeks
+), all_periods as (
+
+    select distinct
+        schedule_id,
+        holiday_week_start as period_start,
+        holiday_week_end as period_end,
+        start_time_utc,
+        end_time_utc,
+        holiday_name_check,
+        true as is_holiday_week
+    from valid_adjustment
+    where holiday_week_start is not null
+        and holiday_week_end is not null
+
+    union all
+
+    select distinct
+        schedule_id,
+        valid_from as period_start,
+        valid_until as period_end,
+        start_time_utc,
+        end_time_utc,
+        cast(null as {{ dbt.type_string() }}) as holiday_name_check,
+        false as is_holiday_week
+    from valid_adjustment
+
+-- We have holiday and non holiday schedules together, now let's sort them to understand the previous end and next start of neighboring schedules
+), sorted_periods as (
+
+    select distinct
+        *,
+        lag(period_end) over (partition by schedule_id order by period_start, start_time_utc) as prev_end,
+        lead(period_start) over (partition by schedule_id order by period_start, start_time_utc) as next_start
+    from all_periods
+
+-- We need to adjust some non holiday schedules in order to properly fill holiday gaps in the schedules later down the transformation
+), non_holiday_period_adjustments as (
+
+    select
+        schedule_id, 
+        period_start, 
+        period_end,
+        prev_end,
+        next_start,
+        coalesce(case 
+            when not is_holiday_week and prev_end is not null then first_value(prev_end) over (partition by schedule_id, period_start order by start_time_utc rows between unbounded preceding and unbounded following)
```

Comment:
Just curious why the unbounded following addition? Do we need to make a similar change in line 192?

Reply:
Great catch and this was a key finding I found in this latest update. Essentially the reason for this change is the following:

- Updated window function
```sql
first_value(prev_end) over (partition by schedule_id, period_start order by start_time_utc rows between unbounded preceding and unbounded following)
```

In this function, for each schedule_id, period_start group, the window is set to span the entire partition (from the first to the last row). The first_value() function then always returns the value of prev_end for the first row in the partition, regardless of where the current row is. This will be the same for all rows in the partition since rows between unbounded preceding and unbounded following looks at the entire partition.

- Previous version of the window function
```sql
first_value(prev_end) over (partition by schedule_id, period_start order by start_time_utc rows unbounded preceding)
```

In this function, for each schedule_id, period_start group, the window is set to start from the first row in the partition and end with the current row, based on the ordering by start_time_utc. Therefore, first_value() will return the value of prev_end for the first row in the window, which is the first row from the start of the partition up to the current row. Unlike the first function, this may yield different results for different rows since the window ends at the current row.

So the difference is in the window of rows they consider for each calculation: the updated window function considers all rows in the partition, while the second one only considers rows from the start up to the current row. I was noticing the different results within the window was causing inconsistencies for these first_value and last_value window function. Therefore, this update ensure the rows within the partition yield the same and accurate results.

Actually, with you brining this up I noticed a few things:
- I actually missed making this update [here](https://github.com/fivetran/dbt_zendesk/blob/215cd87df98f30b75c0005a01c13bdf19317aa35/models/intermediate/int_zendesk__schedule_spine.sql#L300) and just applied it in the latest commit.
- Thanks for bringing up the question around why this was not applied to the `max()` window functions. This was actually not applied as the complexity of the window function was not needed. Since we are just grabbing the max, the extra logic to order and frame the partition was just extra compute that was not necessary. As such it was omitted.
  - In addressing this comment, I actually realize the ordering for the `max()` window functions is completely unnecessary. So I removed the order by and frame clause to reduce the compute of this model as they are not needed.
- Circling back on the code you called out on line 192... I was doing an assessment of this field and realized I never actually ended up using it in the downstream ctes 😱. This must have been an artifact of my development. As such, I was able to remove this cte. Thanks for bringing attention to these as I was able to see a significant improvement in runtime following these updates where I removed erroneous ctes and window function order by and frame clauses.

Reply:
Awesome, thanks for the details Joe! Glad you saw other opportunities to save on compute and runtime too 👍

Reply:
Awesome, thanks for the details Joe! Glad you saw other opportunities to save on compute and runtime too 👍